### PR TITLE
Fix infinite loop snoble found

### PR DESCRIPTION
**Pull request by johnynek**
Never let it be said that I let bugs in totality checking linger!

https://github.com/johnynek/bosatsu/pull/168#issuecomment-472724055

@snoble noticed that we are missing a requirement on recursion: that you cannot recurse on different variables. I added the test he suggested (which did infinitely loop for me when I checked it in Eval, and the test initially failed to disallow).

I'm very grateful to you @snoble for thinking about this, and noticing this problem.

### CODE REVIEW

File: core/src/test/scala/org/bykn/bosatsu/DefRecursionCheckTest.scala [hunk 1]
```diff
@@ -253,7 +270,7 @@ recursive def foo(lst):
   a.plus(b)
 """)
 
-    allowed("""#
+    disallowed("""#
```
Comment (johnynek) [inline core/src/test/scala/org/bykn/bosatsu/DefRecursionCheckTest.scala:294]: @snoble note: your example is no longer legal, which I think is what we want.

**Comment by johnynek** [comment]
@snoble question: should we just limit a single recur match? I don't see why you ever need two. If you are matching on a single variable. Maybe that rule is simpler since it subsumes that all recurs must be on the same variable.

I'm wary the more I think of it, of allowing a pattern that seems unnecessary and encourages writing non-tailrecursive functions.

**Comment by johnynek** [comment]
The more i think, the more I think requiring exactly 1 recur, and requiring at least one branch to make a recursive call is the way to go.

You can always weaken rules later if safe, it is better to start with stricter rules.

**Comment by snoble** [comment]
Yeah, the only reason to have it now (that i can think of) is it might help with how you organize your cute, and it's hard to think of a good use of that

**Comment by snoble** [comment]
Great to see this fix get in so quickly!!

**Pull request merged**